docs/internal/types.md

```diff
@@ -0,0 +1,359 @@
+Specifically, you cannot convert `mut` to `imm` (as you would with a `const` cast in C++), or vice versa.
+A `freeze` operation consumes an `iso`, destroying the only copy of the capability that is required to mutate any object in a region, and returns an `imm` capability.
+There is no corresponding `unfreeze` operation to consume an `imm` capability and provide any kind of mutable capability.
+Most importantly, capabilities are not type traits like C++'s `const int * const`, they have specific monotonic semantics that change the set of operations that are permitted with the resulting capability.
+
+## Module / Class model
+
+TODO
+
+## Classes and Interfaces
+
+Classes and interfaces have a similar structure:
+ * Fields: elements of a specific type that have a runtime space allocated.
+ * Methods: functions that belong to a type.
+
+The declaration is trivial:
+```ts
+// Helps receivers know they can call a function named increment
+interface Example
+{
+  // All derived classes will have the common field
+  var common : I64;
+
+  // Classes that implement this interface must implement this method and provide the field
+  // Classes that inherit from this interface must implement this method but get the field from the base class
+  increment();
+};
+
+// This class inherits, so it gets the field `common` from the interface
+class MyType : Example
+{
+  let field : I32;
+
+  // Some other method
+  innerSum(x : I32) : I32
+  {
+    let res : I32 = x + field;
+    res
+  }
+
+  // Implements `increment`
+  increment()
+  {
+    self.common += 1;
+  }
+};
+
+// This class implements the interface, so it must provide the `common` field
+class OtherType {
+  // Must provide the field
+  var common : I64;
+
+  // Implements `increment`
+  increment()
+  {
+    self.common += 2;
+  }
+};
+```
+
+### Fields
+
+Define objects that belong to the class or interface and will be instantiated (ie. use memory) on every instance of the class.
+
+Fields can be initialised in two ways:
+ * `var`: Must be initialised on construction, but can change after.
+ * `let`: Must be initialised on construction, and cannot change later, like const fields in C++.
+
+They can also have a keyword `embed`, which means they're placed in the struct directly, as opposed to the default of being a pointer to the data.
+
+For example
+```ts
+// This struct layout would be { F64, F64 }
+class Point
+{
+  embed var x : F64;
+  embed var y : F64;
+};
+
+// This struct layout would be { Ref[Point], Ref[Point] }, same as { USize, USize }
+// With `Ref` being the notion of the "address of" an object.
+class Line
+{
+  var orig : Point;
+  var dest : Point;
+};
+```
+
+### Methods
+
+Methods are functions that belong to a class or interface which can have arguments and a return type.
+
+Methods can be called statically (via `Class::method()`) if the compiler can infer the concrete type, or dynamically (via `object.method()`) otherwise.
```

Comment:
I'm confused by "if the compiler can infer the concrete type"; ITYM "`Class::method(object, ...)` can be called if `object` is of type `Class`", though as you point out below, `Class::method()`s need not be given a `self` reference.

Reply:
Ah, I see what you mean. The compiler doesn't need to infer the concrete type in `Class::method()`. I'll update this to include all 4 cases:
```
// Static
Class::method();
obj.method(); // with obj of a known concrete type -> Class::method(obj).

// Dynamic
Interface::method(obj); // which calls obj's concrete class' method
obj.method();
```